### ISSUE TITLE
Writing a simple parser for web crawler output file

### ISSUE THREAD

**Issue by awa5114** [lark-parser/lark#839]
I am interested in parsing typical output of a [website crawler](https://github.com/ahadsheriff/map-website) using `lark`. Here is an example of some sample output based on my own github website:

```
--------------------------------------------------------------------
All found URLs:
https://awa5114.github.io/2021/01/12/#step-2-select-location-and-environment-for-new-project
https://awa5114.github.io/2021/01/12/mypy-pycharm.html
https://awa5114.github.io/2021/01/12/#step-6-test-the-template-on-a-script
--------------------------------------------------------------------
All local URLs:
https://awa5114.github.io/2021/01/12/#step-2-select-location-and-environment-for-new-project
https://awa5114.github.io/2021/01/12/mypy-pycharm.html
--------------------------------------------------------------------
All foreign URLs:
https://github.com/awa5114
https://github.com/jekyll/jekyll
https://github.com/jekyll/minima
--------------------------------------------------------------------
All broken URLs:
```

After following a few `lark` tutorials and reading up on EBNF here's the grammar I came up with:

```
start: ("--------------------------------------------------------------------\nAll " descriptor " URLs:\n" link_list)~4
link_list: (string "\n")*
descriptor: "found" | "local" | "foreign"
string: ESCAPED_STRING
%import common.ESCAPED_STRING
```

I saved this to a file called `grammar.lark` in my current working directory. I then tried to parse 

```
from lark import Lark

if __name__ == '__main__':

    json_parser = Lark.open("grammar.lark")

    with open("test_main_ref.txt") as fo:
        tree = json_parser.parse(fo.read())
```

`test_main_ref.txt` contains the sample output from the crawler shown above. Upon running this script I get the following error:

```
Traceback (most recent call last):
  File "C:/Users/map-website/own_parser.py", line 30, in <module>
    tree = json_parser.parse(fo.read())
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\lark.py", line 495, in parse
    return self.parser.parse(text, start=start)
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\parser_frontends.py", line 223, in parse
    return self._parse(start, text)
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\parser_frontends.py", line 72, in _parse
    return self.parser.parse(input, start, *args)
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\parsers\earley.py", line 294, in parse
    to_scan = self._parse(stream, columns, to_scan, start_symbol)
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\parsers\xearley.py", line 139, in _parse
    to_scan = scan(i, to_scan)
  File "C:\Users\miniconda3\envs\map-website\lib\site-packages\lark\parsers\xearley.py", line 116, in scan
    raise UnexpectedCharacters(stream, i, text_line, text_column, {item.expect.name for item in to_scan}, set(to_scan))
lark.exceptions.UnexpectedCharacters: No terminal defined for 'h' at line 3 col 1

https://awa5114.github.io/2021/01/12/#st
^

Expecting: {'ESCAPED_STRING', '__ANON_0'}

```
I think the error is related to the fact that escaped strings do not contain punctuation. When I was writing the grammar it was my intention that `ESCAPED_STRING` would match the entire url but I now realize that is not the case. I have a couple of questions:

 - Is there another standard terminal in `lark` that I can use to match all characters in a given line including punctuation?
 - How can I further improve the proposed grammar? Am I doing it right?
 - Is the way that I am handling new line characters (i.e by directly matching them) OK?

Thanks in advance for your help.

**Comment by MegaIng** [comment]
> Is there another standard terminal in lark that I can use to match all characters in a given line including punctuation?

No. `ESCAPED_STRING` is meant for complex python-like string literals. for you `/.+/` will probably work and if that doesn't `/\S+/` should work.

> How can I further improve the proposed grammar? Am I doing it right?

This isn't really a typical usecase for a full grammar. Just a linebased iterator could easily deal with this without the need of a full blown parser library

> Is the way that I am handling new line characters (i.e by directly matching them) OK?

Totally. Python and Lark should handle any potential problems with `\r\n` etc.

**Comment by awa5114** [comment]
Thanks @MegaIng. I guess then that Regex is the preferred method to make matches which do not exist in `common`? I changed my `.lark` file to the following and it worked:

```
start: ("--------------------------------------------------------------------\nAll " descriptor " URLs:\n" link_list)~4
link_list: (url "\n")*
descriptor: "found" | "local" | "foreign" | "broken"
url: /.+/
```
I'm getting the result I need, however, I still have a funny feeling like the grammar is not 100 percent optimal. The `start` rule seems a bit too wordy and I am sure there is a more conventional/compact way of dealing with newline characters. 

Finally, I see what you mean by this possibly being an overkill. The grammar for this particular file is not very complex and a line based iterator would have been just fine. But do know that I have spent years writing crappy line iteration based parsers for text files which have much more complex grammars. This is my main incentive for wanting to learn more about Lark and how to come up with good, concise grammars...

**Comment by erezsh** [comment]
For one thing, you can replace `"----"` with `/-{4}/` (regex syntax).

Secondly, it's better to refactor repeated expressions into a rule:

```lark
start: section~4`
section: ...
```

You can use `common.NEWLINE`, which is a little more clever. 

Other than that, it's a very simple file. Not much to say.

**Comment by awa5114** [comment]
Thanks @erezsh, this has helped me improve the grammar to the following:

```
start: section~4
section: (bar  "All " descriptor " URLs:"  link_list)
link_list: (url)*
descriptor: "found" | "local" | "foreign" | "broken"
url: /.+/
bar: /-{68}/

%import common.NEWLINE
%ignore NEWLINE
```

Calling `pretty` on the resulting tree results in the following:

```
start
  section
    bar	--------------------------------------------------------------------
    descriptor
    link_list
      url	https://awa5114.github.io/2021/01/12/#step-2-select-location-and-environment-for-new-project
      url	https://awa5114.github.io/2021/01/12/mypy-pycharm.html
      url	https://awa5114.github.io/2021/01/12/#step-6-test-the-template-on-a-script
  section
    bar	--------------------------------------------------------------------
    descriptor
    link_list
      url	https://awa5114.github.io/2021/01/12/#step-2-select-location-and-environment-for-new-project
      url	https://awa5114.github.io/2021/01/12/mypy-pycharm.html
  section
    bar	--------------------------------------------------------------------
    descriptor
    link_list
      url	https://github.com/awa5114
      url	https://github.com/jekyll/jekyll
      url	https://github.com/jekyll/minima
  section
    bar	--------------------------------------------------------------------
    descriptor
    link_list
```

This looks allright, but I would like to not include the terminal `bar` in my tree. How can I achieve this? I had a look through the [docs ](https://lark-parser.readthedocs.io/en/latest/tree_construction.html) and tried preceding `bar` with an underscore and or question mark, but for some reason that does not help...

Thanks
